[WGSL] Fix hardcoded buffer index calculation in MetalFunctionWriter

tadeuzagallo opened the pull request:
#### 536fc726cd22a49ccfdab8468c47e70e16aed264
<pre>
[WGSL] Fix hardcoded buffer index calculation in MetalFunctionWriter
<a href="https://bugs.webkit.org/show_bug.cgi?id=251264">https://bugs.webkit.org/show_bug.cgi?id=251264</a>
&lt;rdar://problem/104742151&gt;

Reviewed by Myles C. Maxfield.

Add a configuration object to be passed into the WGSL compiler, which for now
only contains the limit for `maxBuffersPlusVertexBuffersForVertexStage`. Also
moved the implementation of the function that calculates the buffer index for
a bind group using the limit into WGSL so it can be shared between the compiler
and the API. The plumbing to get it into the `AST::ShaderModule` is a bit
unfortunate for now, but will be addressed in a subsequent patch.

* Source/WebGPU/WGSL/API.h: Added.
(WGSL::vertexBufferIndexForBindGroup):
* Source/WebGPU/WGSL/AST/ASTShaderModule.h:
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::FunctionDefinitionWriter):
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::emitMetalFunctions):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::parse):
(WGSL::parseLChar):
(WGSL::parseUChar):
(WGSL::Parser&lt;Lexer&gt;::parseShader):
* Source/WebGPU/WGSL/Parser.h:
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::staticCheck):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::vertexBufferIndexForBindGroup const):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::Device::createShaderModule):

Canonical link: <a href="https://commits.webkit.org/259547@main">https://commits.webkit.org/259547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2465e7470c4947f6c28b13304501c0fec396f41e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105024 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14103 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114287 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174470 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5025 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97343 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113300 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110780 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11778 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94779 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93647 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26406 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7441 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27765 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7535 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4346 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13590 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47317 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6570 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9324 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/Parser.h:
```diff
@@ -27,12 +27,13 @@
-Expected<AST::ShaderModule, Error> parseUChar(const String& wgsl);
+struct Configuration;
+
+Expected<AST::ShaderModule, Error> parseLChar(const String& wgsl, const Configuration&);
+Expected<AST::ShaderModule, Error> parseUChar(const String& wgsl, const Configuration&);
```
The _parser_ needs the configuration? Really?

I would expect _parsing_ to not be sensitive to the device we're running on...

tadeuzagallo replied:
It doesn't, I added a FIXME about that. We need a new data structure that will hold both the AST and the configuration (and potentially more data in the future). I'll fix it soon, but for now I just put it in the module.

tadeuzagallo committed:
[WGSL] Fix hardcoded buffer index calculation in MetalFunctionWriter
https://bugs.webkit.org/show_bug.cgi?id=251264
<rdar://problem/104742151>

Reviewed by Myles C. Maxfield.

Add a configuration object to be passed into the WGSL compiler, which for now
only contains the limit for `maxBuffersPlusVertexBuffersForVertexStage`. Also
moved the implementation of the function that calculates the buffer index for
a bind group using the limit into WGSL so it can be shared between the compiler
and the API. The plumbing to get it into the `AST::ShaderModule` is a bit
unfortunate for now, but will be addressed in a subsequent patch.

* Source/WebGPU/WGSL/API.h: Added.
(WGSL::vertexBufferIndexForBindGroup):
* Source/WebGPU/WGSL/AST/ASTShaderModule.h:
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::FunctionDefinitionWriter):
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::emitMetalFunctions):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::parse):
(WGSL::parseLChar):
(WGSL::parseUChar):
(WGSL::Parser<Lexer>::parseShader):
* Source/WebGPU/WGSL/Parser.h:
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::staticCheck):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::vertexBufferIndexForBindGroup const):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::Device::createShaderModule):

Canonical link: https://commits.webkit.org/259547@main
